README.txt
ProtectedDesktops: A program that uses native Windows features to create a fully sandboxed environment 

on Windows.

Version 0.1

This program is in beta. It has been tested on Windows 7 x64.


This program was written as part of an undergraduate research project presented at the Honors Transfer 

Council of California, 26 Mar. 2016.

Copyright 2016 Joshua DuFault




Application Files:


ProtectedDesktops.exe

Console.exe

KeepAlive.exe

DesktopSwitcher.exe

Settings.ini

CloseSandbox.bat



Directions:


1. The user must be running Windows Vista or later and have 3 user accounts set up as follows:

	Account 1: the main user. This is the account you use at the Windows logon screen. This account 

should be a Standard User Account

	Account 2: an administrator account.

	Account 3: an account to be used with the sandbox. The account could be used to isolate 

potentially dangerous programs or protect sensitive data. It obviously can't be used for both at the 

same time.

All accounts must have a password.

The user must have logged on to all 3 accounts through the main Windows logon screen at least once.

Make user you can a run as for Account 3. Hold down shit and right click on a simple program like 

notepad. Select "Run as different user." Enter the username and password of Account 3. Verify the 

program ran.



2. The user must have a file manager other than Windows Explorer installed in a protected location. For 

example: have Explorer++ installed in C:\Program Files (x86)\


3. The Settings.ini file must be updated before use. It is in the following format:

	Main user

	Sandboxed user

	Path to file manager

The included Settings.ini file has an example.


4. The CloseSandbox.bat file also has to be updated.

	The file can be opened in any text editor such as notepad.

	The part of the line that says "Sandbox" must be changed to the name of Account 3.
	
	For example: taskkill /f /t /fi "USERNAME eq MyOtherUserAccount"


4. All program files should be saved in a protected location such as C:\Program Files (x86)\


5. Run ProtectedDesktops.exe and enter the requested information when prompted.


The sandbox is now set up. Run DesktopSwitcher.exe to switch between the sandbox and the main desktop. 

Run CloseSandbox.bat to close the sandbox by right clicking and selecting "Run as administrator." 

Programs launched in the sandbox have no access to programs launched on the main desktop and vice versa. 

Processes cannot jump to the other desktop. They cannot keylog or launch shatter attacks across the 

desktop barrier. They cannot read or write to the memory of processes on the other side of the sandbox. 

Processes run on the sandbox have a different set of user folders then the main user. The two users 

cannot read or write to each other's files. Otherwise, programs run like normal in the sandbox. 

Restarting the computer reverts all settings adjusted by the program back to normal.



Background:


Antivirus programs block only 45% of malware attacks according to Brian Dye, the senior vice president 

for information security at Symantec in a Wall Street Journal interview. Many security products use 

sandboxes to attempt to provide better protection by isolating potentially dangerous processes or 

protect sensitive information; however, many of the implementations are easily bypassed or cause 

additional security problems. Common sandboxing techniques currently in use and their limitations are 

listed below.


	1. Running programs on the same desktop with reduced privileges using a restricted or 

impersonation token:
	
		Malware can use windows messaging based attacks i.e. shatter attacks to jump to a more 

privileged program.
	
		All programs in the same session can read each others memory by default and steal 

passwords or other sensitive data from memory.


	2. Creating separate desktops to isolate potentially dangerous programs or protect sensitive 

data.

		Desktops are secured by user, not by process. Malware launched by the same user can 

easily jump between desktops.

		Programs that switch to a dimmed desktop the way Windows does when UAC is invoked are 

merely mimicking the Windows secure desktop. Only processes with system privileges have access to that 

desktop. Instead, these programs create a simple alternate desktop that is easily accessible.


	3. Combining numbers 1 and 2, and using a restricted token with a made up SID and adding a deny 

for that SID to the main desktop's access control list to prevent malware from jumping to the main 

desktop.

		Restricted tokens must contain the RESTRICTED SID which gives read access to a users 

files and process memory. Malware can still steal passwords from memory or read sensitive files.


	4. Using Job Objects to prevent malware from accessing other processes.

		Job objects don't prevent global hooks. Keylogging and code injection are still 

possible. Malware can still steal passwords and jump to less restricted processes.

		In Windows 7 and below, Job Objects don't nest. This means if a program is put into a 

Job Object, it can't use its own. Many programs use Job Objects to implement their own internal 

sandboxes.


	5. Using integrity levels to prevent shatter attacks, code injection, and key logging.
	
		Integrity levels block most windows messages to more privileged processes but not all. 

Some messages are still allowed through and can be used to launch a shatter attack.

		A program installed in a protected area can have a request in its manifest to bypass 

this protection entirely. Some trusted programs are frequently taken over by malware e.g. internet 

browsers.

		Programs running at one integrity level have access to any other program running at the 

same or lower integrity level.

		Many applications use integrity levels in internal sandboxes creating a child process at 

a lower integrity level than the parent program to run potentially dangerous code. Putting such an 

application in a general purpose sandbox using integrity levels can damage the application's internal 

sandbox by putting the parent and child processes at the same integrity level.

		Programs not designed to run at a low or untrusted integrity level frequently don't work 

correctly when forced to do so.


	6. Using all of the above with a token so restricted that it has no access to the file system.
	
		Effective from a security point of view but such a program is not able to do anything 

useful on its own. It needs to be built specifically to be used with a more privileged broker to provide 

a gateway to system resources. Individual application sandboxes can be written this way but a program 

not built to run in such a sandbox won't work.



Hypothesis: It's possible to use Windows' built in security features in a way to make a sandbox that 

better protects a user's data.


It is possible to launch a program using an impersonation token as an alternate user on a newly created 

desktop and then set the discretionary access control list (DACL) on the desktops to prevent programs 

from jumping over. This will prevent a program launched by any user from accessing the other desktop. 

Desktops block all window messages and global hooks preventing shatter attacks and keylogging. A process 

doesn't have write access to any program launched by a different user, preventing code injection. Each 

user has its own set of files and can't access another user's files.

Any process, regardless of user, can read the memory of any other process in the same logon session at 

the same integrity level. A malicious program can still steal sensitive data out of memory. Processes 

are protected by DACLs and adding a deny entry in each process blocking the sandboxed user would prevent 

those programs from stealing information out of memory. However, this would mean constantly watching for 

every new process that's created to add the needed denies. An exception is that a desktop similiar to 

the Windows secure desktop can be created to enter passwords. Programs running as administrator should 

be kept to a minimum because they can the most damage if compromised but since the StartSandbox.exe 

process has to be run as admin anyway to set all the DACLs, it can create a desktop that only programs 

running as an administrator can access. This won't protect standard processes.

The key to this sandbox is the default DACL. If no DACL is specified when a new process is launched, 

Windows will create a default DACL based on the parent process's token. After the initial startup, the 

parent process is typically explorer.exe. Explorer.exe consists of Windows Explorer and the Start Menu. 

The default DACL can be modified to add the neccessary deny entries to create a secure sandbox. Then, 

any new program launched will automatically receive the same DACL and any new program that it launches 

will receive the same DACL and so on. The only exception is if a launching program specifies a DACL 

during creation which is typically only done for a reason such as creating an internal application 

sandbox. In this case, the default DACLs are still usually inherited automatically, the launching 

program just adds further restrictions.

The resulting sandbox barrier blocks data theft from memory and files, keylogging, code injection, and 

shatter attacks across the sandbox barrier. Even attempts to shut down another process across the 

barrier are blocked. The clipboard and public user folders "C:\Users\Public" can be used to transfer 

files and data in and out of the sandbox. This sandbox can be used either to contain malicious programs 

or protect sensitive  information. Only a process with admin credentials can see across the barrier. A 

desktop switching program run with admin credentials is neccessary to switch visible desktops. A batch 

file running taskkill as administrator can be used to end all programs in the sandbox. The desktop will 

automatically close when all programs with access to it are closed.

Integrity levels and Job Objects aren't considered because they interfer with commonly used 

applications.



Testing:


The sandbox successfully blocked the following actions across the sandbox barrier:

	1. Reading and writing information in a process' memory.

	2. Reading or writing to the other user's files.

	3. Keylogging.

	4. Windows messages.


All tested programs worked normally in the sandbox. This includes involved programs that set their own 

internal sandboxes.



Remaining vulnerabilities:


	1. This only protects against malware running with user rights. If it gets administrator access 

either through user error or a novel privilege escalation attack, security is compromised.

	2. There is no way to isolate the Windows kernel and no way to protect against a kernel exploit.

	3. Processes on either side of the sandbox access some of the same resources. There are multiple 

writable locations that both sides have access to. It might be possible for malware to leave malicious 

code waiting in a place that a vulnerable program on the other side will access.



Conclusion:


The remaining vulnerabilities are significantly more difficult than merely reading a password out of 

memory. Perfect security is impossible but this sandbox is more effective than several other currently 

used types. It is also lighter on system resources:

	Total application size: less than 1MB.

	CPU use after initial setup: 0.

	Installation: not required.


Microsoft could implement a similiar sandbox in an update to Windows 10. Windows 10 currently offers a 

multiple desktop feature but it is entirely cosmetic and offers no security. This would give a native 

way for users to protect sensitive information, such as banking data, from theft and keep important 

files safe from destructive malware, such as ransomware.


Future research possibilities include setting default DACLs on kernel objects and inherited permissions 

on files and registry keys to isolate them against vulnerability 3.


ChangeLog.txt
Version 0.2


Added Windows 8 & 10 support.

Added small reliability and speed improvements.



Version 0.1:


Original upload


Source/Common/Sandbox.h
/*
ProtectedDesktops:
A program that uses native Windows features to create a fully sandboxed environment on Windows.

This program was written as part of an undergraduate research project
presented at the Honors Transfer Council of California, 26 Mar. 2016.

Copyright 2016 Joshua DuFault


This file is part of ProtectedDesktops.

ProtectedDesktops is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

ProtectedDesktops is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with ProtectedDesktops.  If not, see <http://www.gnu.org/licenses/>.

*/

#ifndef SANDBOX_H
#define SANDBOX_H

// Requires unicode
#ifndef UNICODE
#define UNICODE
#endif
#ifndef _UNICODE
#define _UNICODE
#endif

#include <fstream>
#include <iostream>
#include <string>
#include <vector>

#include <Aclapi.h>
#include <sddl.h>
#include <windows.h>

const std::wstring SANDBOX = L"SandboxedDesktop";
const std::wstring MAIN = L"Default";
const std::wstring ADMIN = L"AdminOnlyDesktop";
const std::wstring WINDOWSTATIONMAIN = L"WinSta0";
const std::wstring CONSOLEAPP = L"Console.exe";
const std::wstring KEEPALIVE = L"KeepAlive.exe";
const std::wstring SETTINGSFILE = L"settings.ini";

bool getSettings(std::wstring &mainUser, std::wstring &sandboxUser, std::wstring &fileManager);

bool createDACL(SECURITY_ATTRIBUTES *newDACL);
bool createDACL(const std::wstring &userToAllow, SECURITY_ATTRIBUTES *denyDACL, const std::wstring &userToDeny, bool isDesktop = false);
bool setDACL(SECURITY_ATTRIBUTES newDACL, HANDLE objectHandle, SE_OBJECT_TYPE objectType);
bool setDefaultDACL(SECURITY_ATTRIBUTES newDACL, HANDLE &tokenHandle);
bool usernameToSIDString(const std::wstring &username, std::wstring &userSIDString);

bool keyboardEchoOff();
bool getCredentials(const std::wstring &user, std::wstring &userCreds);

bool launchProcessOnDesktop(std::wstring process, std::wstring desktop, bool resume);
bool launchFileManager(const std::wstring &mainUser, const std::wstring &sandboxUser, const std::wstring &fileManager, const std::wstring &desktop, bool isExplorer);

bool opener(const std::wstring &sandboxD, SECURITY_ATTRIBUTES* denyDACL, HDESK &createSandbox);
bool opener(const std::wstring &sandboxD, SECURITY_ATTRIBUTES* denyDACL, HWINSTA &createSandbox); // Not currently used, provided for potential future use

std::wstring currentDesktopName();
std::wstring currentWindowStationName(); // Not currently used, included for potential future use
bool switchTo(const std::wstring &newDesktop, const std::wstring &newWindowStation = WINDOWSTATIONMAIN);

BOOL SetPrivilege(LPCTSTR Privilege, BOOL bEnablePrivilege);

#endif


Source/Common/Sandbox.cpp
/*
ProtectedDesktops:
A program that uses native Windows features to create a fully sandboxed environment on Windows.

This program was written as part of an undergraduate research project
presented at the Honors Transfer Council of California, 26 Mar. 2016.

Copyright 2016 Joshua DuFault


This file is part of ProtectedDesktops.

ProtectedDesktops is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

ProtectedDesktops is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with ProtectedDesktops.  If not, see <http://www.gnu.org/licenses/>.

*/


#ifndef WIN32_LEAN_AND_MEAN
#define WIN32_LEAN_AND_MEAN
#endif

#include "Sandbox.h"
using namespace std;

// Opens settings file
// gets primary user, user account & file manager to use for sandbox
// Returns true 
bool getSettings(wstring &mainUser, wstring &sandboxUser, wstring &fileManager)
{
	bool getSettingsWorked = false;

	wifstream fin;
	fin.open(SETTINGSFILE.c_str());

	if (fin)
		getline(fin, mainUser);

	if (fin)
		getline(fin, sandboxUser);

	if (fin)
		getline(fin, fileManager);

	if (fin && mainUser != L"" && sandboxUser != L"" && fileManager != L"")
		getSettingsWorked = true;
	else
		fin.clear();

	fin.close();

	return getSettingsWorked;
}

bool createDACL(SECURITY_ATTRIBUTES *newDACL)
{
	bool createDACLWorked = false;

	wstring userDenySIDString, userallowSIDString;

	// Sets owner to Builtin/Administrators
	// Denies access to userToDeny
	// Gives full control to Builtin/Administrators
	// Gives read & execute access to userToAllow
	wstring newACE;
	newACE = L"D:(A;OICI;GA;;;SY)(A;OICI;GA;;;BA)";

	// createDACLWorked is true iff denyDACL is successfully set
	if (ConvertStringSecurityDescriptorToSecurityDescriptor(newACE.c_str(), SDDL_REVISION_1, &(newDACL->lpSecurityDescriptor), NULL))
		createDACLWorked = true;

	return createDACLWorked;
}

// Adds deny ACE entry for specified user
// Returns true on successfull creation of security descriptor
// Returns false otherwise
bool createDACL(const wstring &userToAllow, SECURITY_ATTRIBUTES *newDACL, const wstring &userToDeny, bool isDesktop)
{
	bool createDACLWorked = false;

	wstring userDenySIDString, userallowSIDString;

	// Must have a successful creation of user to allow or only admin will have access
	// Must either have a successful creation of user to deny or have no user to deny
	// No point in continuing otherwise
	if (!usernameToSIDString(userToAllow, userallowSIDString)
		|| !usernameToSIDString(userToDeny, userDenySIDString))
		return createDACLWorked;

	// Sets owner to Builtin/Administrators
	// Denies access to userToDeny
	// Gives full control to Builtin/Administrators
	// Gives read & execute access to userToAllow
	wstring newACE;
	newACE = L"D:(D;OICI;GA;;;" + userDenySIDString + L")(A;OICI;GA;;;SY)(A;OICI;GA;;;BA)(A;OICI;GA;;;" + userallowSIDString + L")";

	// Gives desktop access to Modern apps
	if (isDesktop)
		newACE += L"(A;OICI;GA;;;S-1-15-2-1)";

	// createDACLWorked is true iff denyDACL is successfully set
	if (ConvertStringSecurityDescriptorToSecurityDescriptor(newACE.c_str(), SDDL_REVISION_1, &(newDACL->lpSecurityDescriptor), NULL))
		createDACLWorked = true;

	return createDACLWorked;
}

// Sets DACL and owner on object with specified handle
// Returns true on successful setting of security info
// Returns false otherwise
bool setDACL(SECURITY_ATTRIBUTES newDACL, HANDLE objectHandle, SE_OBJECT_TYPE objectType)
{
	bool setDACLWorked = false;

	BOOL existsACL = TRUE, isDefaultACL = TRUE;
	PACL DACLtoSet;
	PSID newOwner;
	ConvertStringSidToSid(L"BA", &newOwner); // SID for Builtin/Administrators

	// Sets the bit flag mask
	SECURITY_INFORMATION securityToSet;
	if (objectType == SE_KERNEL_OBJECT)
		securityToSet = OWNER_SECURITY_INFORMATION | DACL_SECURITY_INFORMATION | LABEL_SECURITY_INFORMATION;
	else
		securityToSet = OWNER_SECURITY_INFORMATION | DACL_SECURITY_INFORMATION;

	// Pulls the DACL out of the newDACL structure
	// Sets the DACL of the object with specified handle
	// Function is not bool, this return value means success
	if (GetSecurityDescriptorDacl(newDACL.lpSecurityDescriptor, &existsACL, &DACLtoSet, &isDefaultACL)
		&& SetSecurityInfo(objectHandle, objectType, securityToSet, newOwner, NULL, DACLtoSet, NULL) == ERROR_SUCCESS)
		setDACLWorked = true;

	return setDACLWorked;
}

bool setDefaultDACL(SECURITY_ATTRIBUTES newDACL, HANDLE &tokenHandle)
{
	bool setDefaultDACLworked = false;

	PACL sandboxACLtoSet;
	BOOL existsACL = TRUE, isDefaultACL = TRUE;

	if (GetSecurityDescriptorDacl(newDACL.lpSecurityDescriptor, &existsACL, &sandboxACLtoSet, &isDefaultACL))
	{
		TOKEN_DEFAULT_DACL defaultDACL;
		defaultDACL.DefaultDacl = sandboxACLtoSet;

		if (SetTokenInformation(tokenHandle, TokenDefaultDacl, &defaultDACL, sizeof(defaultDACL)))
			setDefaultDACLworked = true;
	}

	return setDefaultDACLworked;
}

// Creates DACL that only gives access to admin and system
// Sets userSIDString to SID wstring version of username
// Returns true on successful completion of ConvertSidToStringSid(), false otherwise
bool usernameToSIDString(const wstring &username, wstring &userSIDString)
{
	bool usernameToSIDStringWorked = false;

	SID userSID[SECURITY_MAX_SID_SIZE];
	DWORD userSIDSize = SECURITY_MAX_SID_SIZE;
	DWORD domainNameSize = 0;
	SID_NAME_USE userType;
	LPTSTR LPuserSIDString = nullptr;

	// Get size of buffer needed for domainName
	LookupAccountName(NULL, username.c_str(), &userSID, &userSIDSize, NULL, &domainNameSize, &userType);
	TCHAR *domainName = new TCHAR[domainNameSize];

	// Create wstring version of SIDS for userToDenySID and userToAllowSID
	if (LookupAccountName(NULL, username.c_str(), &userSID, &userSIDSize, domainName, &domainNameSize, &userType)
		&& ConvertSidToStringSid(&userSID, &LPuserSIDString))
	{
		userSIDString = LPuserSIDString;
		usernameToSIDStringWorked = true;
		// Needed after ConvertSidToStringSid()
		LocalFree(LPuserSIDString);
	}

	delete[] domainName;

	return usernameToSIDStringWorked;
}

// Don't want to echo back password
bool keyboardEchoOff()
{
	bool keyboardEchoOff = false;

	HANDLE consoleInputHandle = GetStdHandle(STD_INPUT_HANDLE);
	DWORD mode;

	if (GetConsoleMode(consoleInputHandle, &mode) && SetConsoleMode(consoleInputHandle, mode - ENABLE_ECHO_INPUT))
		keyboardEchoOff = true;

	return keyboardEchoOff;
}

bool getCredentials(const wstring &user, wstring &userCreds)
{
	bool getCredentials = false;

	// Need credentials for sandboxed user
	wstring sandboxUserCred;
	wcout << L"Sandbox switched to secure desktop to enter credentials." << '\n' << endl;
	wcout << L"Enter password for " << user << ": ";
	if (getline(wcin, userCreds))
		getCredentials = true;
	wcout << '\n' << endl;

	return getCredentials;
}

bool launchProcessOnDesktop(wstring process, wstring desktop, bool resume)
{
	bool launchProcessOnDesktop = false;

	if (!switchTo(desktop))
		return launchProcessOnDesktop;

	// Create new process on admin only desktop to get console there
	PROCESS_INFORMATION processInfo = { 0 };
	STARTUPINFOW processSettings = { 0 };
	processSettings.cb = sizeof(STARTUPINFOW);

	vector<WCHAR> newDesktop(desktop.begin(), desktop.end());
	newDesktop.push_back(0);
	processSettings.lpDesktop = &newDesktop[0];

	if (CreateProcess(process.c_str(), NULL, NULL, NULL, FALSE, CREATE_NEW_CONSOLE | CREATE_SUSPENDED, NULL, NULL, &processSettings, &processInfo))
		launchProcessOnDesktop = true;

	if (resume)
		ResumeThread(processInfo.hThread);

	// Give console.exe time to fully start up
	WaitForSingleObject(processInfo.hProcess, 60000);

	return launchProcessOnDesktop;
}

// Function that launches file manager with impersonation token
// Returns true on successful completion
// Returns false otherwise
bool launchFileManager(const wstring &userToAllow, const wstring &userToDeny, const wstring &fileManager, const wstring &desktop, bool isExplorer)
{
	bool launchFileManagerWorked = false;

	wstring sandboxUserCred;

	// Switch to admin only desktop to get password
	// Get a console on new desktop
	// Get password
	switchTo(ADMIN);
	if (!getCredentials(userToAllow, sandboxUserCred))
		return launchFileManagerWorked;

	// Unused Code to set desktop in STARTUPINFOW, here for documentation
	// Undocumented feature of CreateProcessWithLogonW() causes it to assing new logon SID
	//	when desktop is specified, causing GUI problems
	//	vector<WCHAR> newDesktop(desktop.begin(), desktop.end());
	//	newDesktop.push_back(0);
	//	processSettings.lpDesktop = &newDesktop[0];

	// Switch to indicated desktop to launch program
	switchTo(desktop);

	// By the power of Grayskull . . . I HAVE THE POWER!!!
	SetPrivilege(SE_DEBUG_NAME, TRUE);

	// Start Launcher as new user to get token
	// Using CreateProcessWithLogonW() to retain Logon SID
	PROCESS_INFORMATION launcherInfo = { 0 };
	STARTUPINFOW launcherSettings = { 0 };
	launcherSettings.cb = sizeof(STARTUPINFOW);
	wstring toStart;

	// Creating new process with CreateProcessWithLogonW() to get impersonation of specified user with old Logon SID
	if (isExplorer)
	{
		toStart = L"C:\\Windows\\Explorer.exe";

		// Get a handle to explorer.exe
		// Return on failure
		DWORD explorerID;
		HWND explorerWindow = FindWindow(L"Shell_TrayWnd", NULL);
		if (explorerWindow == NULL)
			return launchFileManagerWorked;
		GetWindowThreadProcessId(explorerWindow, &explorerID);
		HANDLE explorerHANDLE = OpenProcess(PROCESS_ALL_ACCESS, FALSE, explorerID);
		if (explorerHANDLE == NULL)
			return launchFileManagerWorked;

		// Try to close Explorer.exe using Window message exit code
		//	If that fails, terminate process
		if (!PostMessage(explorerWindow, WM_USER + 436, 0, 0))
			TerminateProcess(explorerHANDLE, 0);

		// Give shell time to exit fully
		// WaitForSingleObject() isn't working correctly with explorer.exe on Windows 10
		DWORD exitCode;
		int count = 0;
		do
		{
			Sleep(500);
			GetExitCodeProcess(explorerHANDLE, &exitCode);
			count++;
		} while (exitCode == STILL_ACTIVE && count < 10);

		if (!CreateProcessWithLogonW(userToAllow.c_str(), NULL, sandboxUserCred.c_str(), LOGON_WITH_PROFILE, KEEPALIVE.c_str(), NULL, CREATE_NEW_CONSOLE, NULL, NULL, &launcherSettings, &launcherInfo))
			return launchFileManagerWorked;
	}
	else
	{
		toStart = fileManager;

		if (!CreateProcessWithLogonW(userToAllow.c_str(), NULL, sandboxUserCred.c_str(), LOGON_WITH_PROFILE, KEEPALIVE.c_str(), NULL, CREATE_NEW_CONSOLE, NULL, NULL, &launcherSettings, &launcherInfo))
			return launchFileManagerWorked;
	}

	// Gives the keepAlive program time to start fully
	WaitForSingleObject(launcherInfo.hProcess, 5000);

	// Code to impersonate using LogonUser() left for documentation
	// Creates new Logon SID
	// LogonUser(userToAllow.c_str(), NULL, sandboxUserCred.c_str(), LOGON32_LOGON_INTERACTIVE, LOGON32_PROVIDER_DEFAULT, &tokenHandle)

	// Clear password as soon as possible
	SecureZeroMemory(&sandboxUserCred, sandboxUserCred.size());

	PROCESS_INFORMATION processInfo = { 0 };
	STARTUPINFOW processSettings = { 0 };
	processSettings.cb = sizeof(STARTUPINFOW);
	SECURITY_ATTRIBUTES processDACL;
	HANDLE processTokenHandle = { 0 };
	HANDLE tokenHandle = { 0 };

	// Taking token from CreateProcessWithLogonW(), duplicating it, setting new default DACL and using that to launch file manager
	if (createDACL(userToAllow, &processDACL, userToDeny)
		&& OpenProcessToken(launcherInfo.hProcess, MAXIMUM_ALLOWED, &processTokenHandle)
		&& DuplicateTokenEx(processTokenHandle, MAXIMUM_ALLOWED, &processDACL, SecurityDelegation, TokenPrimary, &tokenHandle)
		&& setDefaultDACL(processDACL, tokenHandle)
		&& CreateProcessWithTokenW(tokenHandle, NULL, toStart.c_str(), NULL, 0, 0, NULL, &processSettings, &processInfo))
		launchFileManagerWorked = true;

	// Let the power return.
	SetPrivilege(SE_DEBUG_NAME, FALSE);

	// Clean up
//	ResumeThread(launcherInfo.hThread);
	LocalFree(processDACL.lpSecurityDescriptor);
	CloseHandle(launcherInfo.hThread);
	CloseHandle(launcherInfo.hProcess);
	CloseHandle(tokenHandle);

	// These were only created if everything worked
	if (launchFileManagerWorked)
	{
		CloseHandle(processInfo.hThread);
		CloseHandle(processInfo.hProcess);
	}

	return launchFileManagerWorked;
}

// Takes name of desktop to be created and security attribute structure to be applied to desktop
// Returns true if creation successful
// Returns false otherwise
bool opener(const wstring &sandboxD, SECURITY_ATTRIBUTES* denyDACL, HDESK &createSandbox)
{
	bool openerWorked = false;

	// true iff a handle to sandboxD desktop is returned
	createSandbox = CreateDesktop(sandboxD.c_str(), NULL, NULL, 0, GENERIC_ALL, denyDACL);

	if (createSandbox != NULL)
		openerWorked = true;

	return openerWorked;
}

// Takes name of desktop to be created and security attribute structure to be applied to desktop
// Returns true if creation successful
// Returns false otherwise
bool opener(const wstring &windowStation, SECURITY_ATTRIBUTES* allowDACL, HWINSTA &windowStationHandle)
{
	bool openerWorked = false;

	// true iff a handle to sandboxD desktop is returned
	windowStationHandle = CreateWindowStation(windowStation.c_str(), NULL, GENERIC_ALL, allowDACL);

	if (windowStationHandle != NULL && SetProcessWindowStation(windowStationHandle))
		openerWorked = true;

	return openerWorked;
}

// Returns name of currently active desktop
wstring currentDesktopName()
{

	HDESK currentDesktop = OpenInputDesktop(0, FALSE, DESKTOP_SWITCHDESKTOP);
	wstring currentNameString;

	// if currentDesktop handle is valid, get name of currentDesktop
	if (currentDesktop != NULL)
	{
		const int OBJECTNAME = 2; // Tells GetUserObjectInformation() to return name of desktop
		DWORD sizeNeeded = 0;
		// Get size of buffer needed
		GetUserObjectInformation(currentDesktop, OBJECTNAME, NULL, NULL, &sizeNeeded);

		// Create buffer of correct size
		TCHAR *currentName = new TCHAR[sizeNeeded];
		// Get name
		if (GetUserObjectInformation(currentDesktop, OBJECTNAME, currentName, sizeNeeded, NULL))
			currentNameString = currentName;

		delete[] currentName;

		CloseDesktop(currentDesktop);
	}

	return currentNameString;
}

// Returns name of currently window station
wstring currentWindowStationName()
{

	HWINSTA currentWindowStation = GetProcessWindowStation();

	wstring currentNameString;

	// if currentDesktop handle is valid, get name of currentDesktop
	if (currentWindowStation != NULL)
	{
		const int OBJECTNAME = 2; // Tells GetUserObjectInformation() to return name of desktop
		DWORD sizeNeeded = 0;
		// Get size of buffer needed
		GetUserObjectInformation(currentWindowStation, OBJECTNAME, NULL, NULL, &sizeNeeded);

		// Create buffer of correct size
		TCHAR *currentName = new TCHAR[sizeNeeded];
		// Get name
		if (GetUserObjectInformation(currentWindowStation, OBJECTNAME, currentName, sizeNeeded, NULL))
			currentNameString = currentName;

		delete[] currentName;

		CloseWindowStation(currentWindowStation);
	}

	return currentNameString;
}

// Switches process and display to indicated destkop
bool switchTo(const wstring &newDesktop, const wstring &newWindowStation)
{
	bool switchToWorked = false;

	// Gets handle to workstation/desktop to switch to
	HWINSTA windowStationHandle = OpenWindowStation(newWindowStation.c_str(), FALSE, GENERIC_ALL);
	if (!SetProcessWindowStation(windowStationHandle))
		return switchToWorked;
	else
		CloseWindowStation(windowStationHandle);

	HDESK newDesktopHandle = OpenDesktop(newDesktop.c_str(), 0, FALSE, GENERIC_ALL);

	// switchToWorked is true iff SwitchDesktop worked
	if (newDesktopHandle != NULL && SetThreadDesktop(newDesktopHandle) && SwitchDesktop(newDesktopHandle))
	{
		CloseDesktop(newDesktopHandle);
		switchToWorked = true;
	}

	return switchToWorked;
}

// Sets and removes SE_DEBUG privilege
// Code mostly straight from MSDN example
BOOL SetPrivilege(
	LPCTSTR Privilege,  // Privilege to enable/disable 
	BOOL bEnablePrivilege  // TRUE to enable. FALSE to disable 
	)
{
	HANDLE hToken;
	OpenProcessToken(GetCurrentProcess(), TOKEN_ALL_ACCESS, &hToken);

	TOKEN_PRIVILEGES tp = { 0 };
	// Initialize everything to zero 
	LUID luid;
	DWORD cb = sizeof(TOKEN_PRIVILEGES);
	if (!LookupPrivilegeValue(NULL, Privilege, &luid))
		return FALSE;
	tp.PrivilegeCount = 1;
	tp.Privileges[0].Luid = luid;
	if (bEnablePrivilege) {
		tp.Privileges[0].Attributes = SE_PRIVILEGE_ENABLED;
	}
	else {
		tp.Privileges[0].Attributes = 0;
	}
	AdjustTokenPrivileges(hToken, FALSE, &tp, cb, NULL, NULL);

	CloseHandle(hToken);

	if (GetLastError() != ERROR_SUCCESS)
		return FALSE;

	return TRUE;
}


Source/RunMe/RunMe.cpp
/*
ProtectedDesktops:
A program that uses native Windows features to create a fully sandboxed environment on Windows.

This program was written as part of an undergraduate research project
presented at the Honors Transfer Council of California, 26 Mar. 2016.

Copyright 2016 Joshua DuFault


This file is part of ProtectedDesktops.

ProtectedDesktops is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

ProtectedDesktops is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with ProtectedDesktops.  If not, see <http://www.gnu.org/licenses/>.

*/

#ifndef WIN32_LEAN_AND_MEAN
#define WIN32_LEAN_AND_MEAN
#endif

// Using std MIN and MAX definitions rather than windows.h ones
#ifndef NOMINMAX
#define NOMINMAX
#endif

#include "Sandbox.h"
using namespace std;


// Seperate launcher is a workaround in Windows 8 & 10
// Launching console program creates conhost.exe as a child process which prevents SetThreadDesktop() from working
// Using a launcher keeps conhost.exe attached to the launcher
int wmain()
{
	PROCESS_INFORMATION processInfo = { 0 };
	STARTUPINFOW processSettings = { 0 };
	processSettings.cb = sizeof(STARTUPINFOW);

	int choice = 0;

	cout << "Protected Desktops: A Sandboxing Program" << endl;

	do
	{
		cout << '\n'
			<< "Pick an option:" << '\n'
			<< "1: Create Sandbox" << '\n'
			<< "2: Switch desktops" << '\n'
			<< "3: Close Sandbox (forcefully closes any open programs)" << '\n'
			<< "4: Exit" << '\n'
			<< "Selection: "
			<< flush;

		cin >> choice;

		if (cin.fail())
		{
			cin.clear();
			choice = 0;
		}

		cin.ignore(numeric_limits<streamsize>::max(), L'\n');

		wstring process;
		if (choice == 1)
			process = L"ProtectedDesktops.exe";
		else if (choice == 2)
			process = L"DesktopSwitcher.exe";
		else if (choice == 3)
			process = L"CloseSandbox.bat";
		else if (choice == 0 || choice == 4)
			continue;


		if (CreateProcess(process.c_str(), NULL, NULL, NULL, FALSE, DETACHED_PROCESS, NULL, NULL, &processSettings, &processInfo))
		{
			cout << "Working..." << endl;
			WaitForSingleObject(processInfo.hProcess, 60000);
		}
		else
			cout << '\n' << "Error. Try again." << '\n';

	} while (choice != 4);

	return 0;
}
